Fixed #31621 -- Add support for '--parallel auto' to test command

adamchainz opened the pull request:
[ticket](https://code.djangoproject.com/ticket/31621#ticket)

adamchainz commented:
@lothemar This PR intersects with your project as well.

lothemar commented on tests/runtests.py:
```diff
@@ -126,11 +126,7 @@ def setup(verbosity, test_labels, parallel, start_at, start_after):
         test_labels_set.add('.'.join(bits))
 
     if verbosity >= 1:
-        msg = "Testing against Django installed in '%s'" % os.path.dirname(django.__file__)
```
Is there a reason behind changing this? It added a nice indicator whether or not the tests were running in parallel.

Non-blocking change though. Other than this, the patch is great.

adamchainz replied:
I removed it during iteration because `default_test_processes()` went away, didn't think to restore it. Have done now!

felixxm commented:
This PR changes the current behavior, e.g. on Jenkins we specify `DJANGO_TEST_PROCESSES=1` and with this PR it runs tests in 9 parallel processes instead of 1.

adamchainz commented:
Right sorry - forgot to copy-paste some stuff from the runner back into `runtests.py`

Imo though, `runtests.py` should default parallel to "auto". Why can't Jenkins invoke it with `--parallel 1` rather than the env var? I'm not even sure why the env var exists - none of the other options can be overridden with env vars.

orf commented on django/test/runner.py:
```diff
@@ -283,17 +284,23 @@ def run(self, test):
+        return int(string)
+    except ValueError:
+        raise argparse.ArgumentTypeError(
+            "{} is not an integer or the string 'auto'".format(repr(string))
```
We could use `{!r}` here to remove the need for manually calling `repr(string)`?

adamchainz replied:
I can even use an f-string since we support Python 3.6+ only now! It looks like it will be the first in the code base though??

orf replied:
Historic moment!

felixxm commented:
@adamchainz Do you have time to keep working on this?

felixxm commented:
Closing due to inactivity.

adamchainz commented:
I've rebased the branch and updated the docs for 4.0.

I believe I already fixed the concern with Jenkins back in the June 2020 push. When I run locally with `DJANGO_TEST_PROCESSES=1 ./runtests.py test_runner --parallel` (plus a patch for macOS to use fork), I get one process.

adamchainz commented:
Oh GitHub doesn't allow reopening a PR after the branch has been force-pushed. Recreated PR here: #14614

felixxm commented:
Thanks
